src/Functions/dateName.cpp

```diff
@@ -0,0 +1,324 @@
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypeDate.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypeDateTime64.h>
+#include <Columns/ColumnString.h>
+
+#include <Functions/DateTimeTransforms.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/FunctionsConversion.h>
+#include <Functions/IFunctionImpl.h>
+#include <Functions/castTypeToEither.h>
+#include <Functions/extractTimeZoneFromFunctionArguments.h>
+
+#include <IO/WriteHelpers.h>
+
+#include <common/DateLUTImpl.h>
+#include <common/find_symbols.h>
+#include <Core/DecimalFunctions.h>
+
+#include <type_traits>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int ILLEGAL_COLUMN;
+    extern const int BAD_ARGUMENTS;
+}
+
+namespace {
+
+template <typename DataType> struct ActionValueTypeMap {};
+template <> struct ActionValueTypeMap<DataTypeDate> { using ActionValueType = UInt16; };
+template <> struct ActionValueTypeMap<DataTypeDateTime>      { using ActionValueType = UInt32; };
+template <> struct ActionValueTypeMap<DataTypeDateTime64>    { using ActionValueType = Int64; };
+
+class FunctionDateNameImpl : public IFunction
+{
+public:
+    static constexpr auto name = "dateName";
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionDateNameImpl>(); }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    bool isVariadic() const override { return true; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (arguments.size() != 2 && arguments.size() != 3)
+            throw Exception(
```

Comment:
Consider to use fmt like Exception throw.
```
throw Exception(
     ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, 
    "Number of arguments for function {} doesn't match: passed {}", 
     getName(), 
     toString(arguments.size())).
```

Reply:
Fixed